pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,225 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: jefftk@google.com (Jeff Kaufman)
+
+#include "pagespeed/system/controller_manager.h"
+
+#include <signal.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <cerrno>
+#include <cstddef>
+#include <cstdlib>
+
+#include "base/logging.h"
+#include "pagespeed/kernel/base/basictypes.h"
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/system/system_rewrite_options.h"
+
+namespace net_instaweb {
+
+
+ControllerManager::ProcessDeathWatcherThread::ProcessDeathWatcherThread(
+    ThreadSystem* thread_system,
+    int controller_read_fd,
+    MessageHandler* handler) : Thread(thread_system,
+                                      "process death watcher",
+                                      ThreadSystem::kDetached),
+                              handler_(handler),
+                              controller_read_fd_(controller_read_fd) {
+}
+
+void ControllerManager::ProcessDeathWatcherThread::Run() {
+  handler_->Message(kInfo, "Watching the root process to exit if it does.");
+
+  ssize_t status;
+  char buf[1];
+  do {
+    status = read(controller_read_fd_, buf, 1);
+  } while (status == -1 && (errno == EINTR ||
+                            errno == EAGAIN));
+  if (status == -1) {
+    handler_->Message(
+        kWarning, "Controller got error %d reading from pipe, shutting down",
+        errno);
+  } else if (status == 0 /* EOF */) {
+    handler_->Message(kInfo, "Root process exited; controller shutting down.");
+  } else if (status == 1 /* read a byte */) {
+    handler_->Message(
+        kInfo, "Root process is starting a new controller; shutting down.");
+  } else {
+    LOG(FATAL) << "Status of " << status << " doesn't make sense";
+  }
+
+  // TODO(jefftk): Once the controller is doing real work and there's real
+  // cleanup work to do, we can initiate that from here.
+
+  exit(EXIT_SUCCESS);
+
+  // Because the controller is exiting via exit(), the babysitter process will
+  // see that and quit itself instead of restarting it.
+}
+
+void ControllerManager::RunController(SystemRewriteDriverFactory* factory,
+                                      MessageHandler* handler) {
+  handler->Message(kInfo, "Controller running with PID %d", getpid());
+
+  // Would set up gRPC server here and pass control to its event loop.  Instead
+  // just hang out sleeping until the ProcessDeathWatcherThread decides it's
+  // time to exit.
+  while (true) {
+    sleep(60);  // 1m
+  }
+}
+
+void ControllerManager::Daemonize(MessageHandler* handler) {
+  // Make a new session (process group).
+  if (setsid() < 0) {
```

Comment:
We fork twice, and this happens between the two forks.

First we fork to get a process we can mess with.  We call setsid in that process to detach from the previous process and be a session group leader.  Then we fork again (and have the parent die) to make sure there's no session group leader.